### ISSUE TITLE
openjdk cannot be installed through bash script using `-u`

### ISSUE THREAD

**Issue by hugobuddel** [conda-forge/openjdk-feedstock#87]
Edit: the issue below was caused by the `-u` flag in bash, see the post below. So my personal immediate problem is solved by simply not using that flag. However, it might be worthwhile to fix the problem in the openjdk feedstock anyway.

---

- [x] I read [the conda-forge documentation](https://conda-forge.org/docs/user/introduction.html#how-can-i-install-packages-from-conda-forge) and could not find the solution for my problem there.

Issue:

It seems not possible to install openjdk through a bash script.

```
$ cat openjdktest.sh
set -Eeuo pipefail
set -x
# one of the below is necessary:
# eval "$(conda shell.bash hook)"
source "$CONDA_PREFIX/etc/profile.d/conda.sh"
conda create -y -n openjdktest
conda activate openjdktest
conda install -c conda-forge -y openjdk

$ bash openjdktest.sh 
...
  zstd               conda-forge/linux-64::zstd-1.5.0-ha95c52a_0


Preparing transaction: done
Verifying transaction: done
Executing transaction: done
+ __conda_reactivate
+ local ask_conda
++ PS1='(openjdktest) '
++ __conda_exe shell.posix reactivate
++ __add_sys_prefix_to_path
++ '[' -n '' ']'
+++ dirname /home/.../bin/conda
++ SYSP=/home/.../bin
+++ dirname /home/.../bin
++ SYSP=/home/...
++ '[' -n '' ']'
++ PATH=/home/.../bin:/home/.../envs/openjdktest/bin:/home/.../condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
++ export PATH
++ /home/.../bin/conda shell.posix reactivate
+ ask_conda='. "/home/.../envs/openjdktest/etc/conda/deactivate.d/openjdk_deactivate.sh"
PS1='\''(openjdktest) '\''
export PATH='\''/home/.../envs/openjdktest/bin:/home/.../condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin'\''
export CONDA_SHLVL='\''2'\''
export CONDA_PROMPT_MODIFIER='\''(openjdktest) '\''
. "/home/.../envs/openjdktest/etc/conda/activate.d/openjdk_activate.sh"'
+ eval '. "/home/.../envs/openjdktest/etc/conda/deactivate.d/openjdk_deactivate.sh"
PS1='\''(openjdktest) '\''
export PATH='\''/home/.../envs/openjdktest/bin:/home/.../condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin'\''
export CONDA_SHLVL='\''2'\''
export CONDA_PROMPT_MODIFIER='\''(openjdktest) '\''
. "/home/.../envs/openjdktest/etc/conda/activate.d/openjdk_activate.sh"'
++ . /home/.../envs/openjdktest/etc/conda/deactivate.d/openjdk_deactivate.sh
/home/.../envs/openjdktest/etc/conda/deactivate.d/openjdk_deactivate.sh: line 1: JAVA_HOME_CONDA_BACKUP: unbound variable
```

That is, it seems that `conda install` calls `__conda_reactivate`, which calls `openjdk_deactivate.sh`, which fails because for some reason `openjdk_activate.sh` is never ran.

Note that performing the commands in ` openjdktest.sh` directly in the shell itself will work. See also https://stackoverflow.com/questions/52779016/conda-command-working-in-command-prompt-but-not-in-bash-script/52813960

Arguably this is not directly a problem caused by openjdk feedstock itself, but rather an inconsistency within conda. ~However, at this moment I do not know how to install openjdk through a bash script. As in, I do not yet have a workaround for this problem, so I'm posting this issue here.~ And it seems that the problem can be solved by having the deactivate script only reset the environment variables if the backup variables actually have been set.

<br/>
Environment (<code>conda list</code>):
<details>

```
$ conda list
# packages in environment at /home/...:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       1_gnu    conda-forge
beautifulsoup4            4.10.0             pyha770c72_0    conda-forge
brotlipy                  0.7.0           py38h497a2fe_1003    conda-forge
bzip2                     1.0.8                h7f98852_4    conda-forge
ca-certificates           2021.10.8            ha878542_0    conda-forge
certifi                   2021.10.8        py38h578d9bd_1    conda-forge
cffi                      1.14.6           py38h400218f_0  
chardet                   4.0.0            py38h578d9bd_2    conda-forge
charset-normalizer        2.0.0              pyhd8ed1ab_0    conda-forge
colorama                  0.4.4              pyh9f0ad1d_0    conda-forge
conda                     4.10.3           py38h578d9bd_4    conda-forge
conda-build               3.21.6           py38h578d9bd_0    conda-forge
conda-package-handling    1.7.3            py38h497a2fe_1    conda-forge
cryptography              35.0.0           py38h3e25421_2    conda-forge
filelock                  3.4.0              pyhd8ed1ab_0    conda-forge
glob2                     0.7                        py_0    conda-forge
gprof2dot                 2019.11.30                 py_0    conda-forge
icu                       69.1                 h9c3ff4c_0    conda-forge
idna                      3.1                pyhd3deb0d_0    conda-forge
jinja2                    3.0.3              pyhd8ed1ab_0    conda-forge
ld_impl_linux-64          2.36.1               hea4e1c9_2    conda-forge
libarchive                3.5.2                hccf745f_1    conda-forge
libffi                    3.3                  h58526e2_2    conda-forge
libgcc-ng                 11.2.0              h1d223b6_11    conda-forge
libgomp                   11.2.0              h1d223b6_11    conda-forge
libiconv                  1.16                 h516909a_0    conda-forge
liblief                   0.11.5               h9c3ff4c_1    conda-forge
libstdcxx-ng              11.2.0              he4da1e4_11    conda-forge
libxml2                   2.9.12               h885dcf4_1    conda-forge
libzlib                   1.2.11            h36c2ea0_1013    conda-forge
lxml                      4.6.2                    pypi_0    pypi
lz4-c                     1.9.3                h9c3ff4c_1    conda-forge
lzo                       2.10              h516909a_1000    conda-forge
markupsafe                2.0.1            py38h497a2fe_1    conda-forge
ncurses                   6.2                  h58526e2_4    conda-forge
openssl                   1.1.1l               h7f98852_0    conda-forge
patchelf                  0.13                 h58526e2_0    conda-forge
pkginfo                   1.8.1              pyhd8ed1ab_0    conda-forge
psutil                    5.8.0            py38h497a2fe_2    conda-forge
py-lief                   0.11.5           py38h709712a_1    conda-forge
pycosat                   0.6.3           py38h497a2fe_1009    conda-forge
pycparser                 2.21               pyhd8ed1ab_0    conda-forge
pyopenssl                 21.0.0             pyhd8ed1ab_0    conda-forge
pysocks                   1.7.1            py38h578d9bd_4    conda-forge
python                    3.8.5                h7579374_1  
python-libarchive-c       3.1              py38h578d9bd_1    conda-forge
python_abi                3.8                      2_cp38    conda-forge
pytz                      2021.3             pyhd8ed1ab_0    conda-forge
pyyaml                    6.0              py38h497a2fe_3    conda-forge
readline                  8.1                  h46c0cb4_0    conda-forge
requests                  2.26.0             pyhd8ed1ab_0    conda-forge
ripgrep                   13.0.0               habb4d0f_1    conda-forge
ruamel_yaml               0.15.80         py38h497a2fe_1006    conda-forge
setuptools                59.2.0           py38h578d9bd_0    conda-forge
six                       1.16.0             pyh6c4a22f_0    conda-forge
soupsieve                 2.3                pyhd8ed1ab_0    conda-forge
sqlite                    3.36.0               h9cd32fc_2    conda-forge
svg2tikz                  1.0.0.dev0               pypi_0    pypi
tk                        8.6.11               h27826a3_1    conda-forge
tqdm                      4.62.3             pyhd8ed1ab_0    conda-forge
urllib3                   1.26.7             pyhd8ed1ab_0    conda-forge
xz                        5.2.5                h516909a_1    conda-forge
yaml                      0.2.5                h516909a_0    conda-forge
zlib                      1.2.11            h36c2ea0_1013    conda-forge
zstd                      1.5.0                ha95c52a_0    conda-forge

```
</details>

<br/>
Details about  <code>conda</code> and system ( <code>conda info</code> ):
<details>

```
$ conda info
     active environment : base
    active env location : /home/.../miniconda3
            shell level : 1
       user config file : /home/.../.condarc
 populated config files : /home/.../.condarc
          conda version : 4.10.3
    conda-build version : 3.21.6
         python version : 3.8.5.final.0
       virtual packages : __cuda=11.2=0
                          __linux=5.4.0=0
                          __glibc=2.31=0
                          __unix=0=0
                          __archspec=1=x86_64
       base environment : /home/.../miniconda3  (writable)
      conda av data dir : /home/.../miniconda3/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/.../miniconda3/pkgs
                          /home/.../.conda/pkgs
       envs directories : /home/.../miniconda3/envs
                          /home/.../.conda/envs
               platform : linux-64
             user-agent : conda/4.10.3 requests/2.26.0 CPython/3.8.5 Linux/5.4.0-90-generic ubuntu/20.04.3 glibc/2.31
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```
</details>

**Comment by hugobuddel** [comment]
Hmm, apparently removing the first line of the bash script makes it work. That is
```
set -Eeuo pipefail
```
causes the problem. This line is from https://betterdev.blog/minimal-safe-bash-script-template/

In particular, it is the `-u` that causes the script to fail. Which is the point https://vaneyckt.io/posts/safer_bash_scripts_with_set_euxo_pipefail/:

> This option causes the bash shell to treat unset variables as an error and exit immediately. Unset variables are a common cause of bugs in shell scripts, so having unset variables cause an immediate exit is often highly desirable behavior.

Nevertheless, there still is a problem that `JAVA_HOME` is not properly reset when `-u` is not used:
```
$ cat openjdktest2.sh 
set -Eeo pipefail

JAVA_HOME=/hello/there
echo "At the start JAVA_HOME is $JAVA_HOME"

source "$CONDA_PREFIX/etc/profile.d/conda.sh"
conda create -y -n openjdktest
conda activate openjdktest
conda install -c conda-forge -y openjdk
conda deactivate

echo "At the end JAVA_HOME is $JAVA_HOME"


$ bash openjdktest2.sh                
At the start JAVA_HOME is /hello/there                                                     
Collecting package metadata (current_repodata.json): done     
Solving environment: done                                                                  

...

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
At the end JAVA_HOME is 
```

So perhaps it is still worthwhile to add some protection clause to the deactivate scripts.

**Comment by hugobuddel** [comment]
I just verified that this is indeed still occurring.

The solution would probably be to check whether `$JAVA_HOME_CONDA_BACKUP` is set at the start of recipe/scripts/deactivate.sh (and  quit if it's not).

I should be able to get you a PR for that.

But so far I have not been able to actually build the openjdk feedstock as-is. Probably due to issues with my own configuration, so I'll give it some more attempts, and let you know if I fail.

(I also realized that your simple reply gave me the confidence that this effort might be worthwhile, even though I myself am not relying on this package anymore. So much better than just closing it due to inactivity, thanks.)

FWIW, this issue helped me come to the conclusion that relying on activate and deactivate scripts is ultimately doomed. So I'm trying to move away from conda, and go the Guix route instead. But nevertheless, it would be good to get this fixed.

**Comment by h-vetinari** [comment]
> But so far I have not been able to actually build the openjdk feedstock as-is. Probably due to issues with my own configuration, so I'll give it some more attempts, and let you know if I fail.

You don't have to build the feedstock from scratch, you could just manually edit the activate/deactivate scripts in an installed environment until it works, and then take those changes to make a PR. FWIW, we recently did a lot of clean-ups for the activation script of arrow, maybe there's some bits that can serve as inspiration (though that one doesn't have a deactivation script, so doesn't have to deal with backing up env vars etc.)

> FWIW, this issue helped me come to the conclusion that relying on activate and deactivate scripts is ultimately doomed.

This is throwing the baby out with the bath water to a large extent. No-one _likes_ having activation scripts, but there are some problems that are just not possible to solve sanely through installation (copying files) alone.

**Comment by hugobuddel** [comment]
> You don't have to build the feedstock from scratch, you could just manually edit the activate/deactivate scripts in an installed environment until it works, and then take those changes to make a PR.

It would probably be possible to develop the fix by just taking the activate and deactivate scripts and run those. That is, ignoring the package itself entirely. 

Nevertheless, I´d like to use the two testcases above to verify the problem is properly resolved. The conda installer plays a role too, because that decides when to call which (de)activation script. So I prefer to do the full process at least once.

Luckily I did manage to get everything compiled! The Python in my base environment was too old, and that is somewhat hard/unsupported to update. (I don't understand how the base environment comes into play at all though.) But then I accidentally installed a broken version of the openjdk package in my base environment that I can't seem to uninstall. And that makes testing hard, because the point is that conda runs the deactivate script without ever having called the activate script. But since openjdk is now in my base environment, it will always call that activate script. So the above tests will not behave as expected.. I'll probably reinstall miniconda or so..

(It should not be so easy to install packages in the base environment... Even with the environment in my prompt, I still occasionally make that mistake, and it can be hard to undo.)

> > FWIW, this issue helped me come to the conclusion that relying on activate and deactivate scripts is ultimately doomed.
> 
> This is throwing the baby out with the bath water to a large extent. No-one _likes_ having activation scripts, but there are some problems that are just not possible to solve sanely through installation (copying files) alone.

Yeah that is perhaps true. The Guix project shows it is not that insane anymore though. The problem with the (de)activation scripts is that every script should behave well, and this issue shows that it is not easy to write them correctly.

I don't know. Maybe conda-forge should add an automated test for each package. After building the package, create a backup of the full shell environment, create an conda environment with the package, activate that environment, deactivate it, compare the resulting shell environment  with the backup. And run that through a shell script with `-u`. Maybe then it could work.

**Comment by hugobuddel** [comment]
There are actually two problems, one with installing openjdk, and one with uninstalling openjdk. And a minor new one when upgrading.

When installing openjdk, conda installs openjdk, then deactivates and (re)activates the environment. If one then manually deactivates the environment, `$JAVA_HOME` would be unset. This problem is fixed in #139. A more proper fix would be that conda first deactivates the environment, then installs the new packages, and then activates the environment, but that is out of our control.

The other, still existing problem, is that if one uninstalls openjdk, and deactivates the environment, then the deactivation script of openjdk is not called, so `$JAVA_HOME` is not restored to its original value, but instead points to a directory that no longer exists. Conda here also does the deactivate-activate cycle, but after the uninstallation of openjdk. A good fix would be conda first deactivating the environment, then uninstalling openjdk, then activating the environment again. But that is once again out of our control.

A third problem could arise with updating the packages. Then conda would update the package, then call deactivate / activate with the new package installed. So then things would go wrong if the scripts of the different versions don´t match. I tried to pay attention to this, but I think that if `$JAVA_HOME` was unset, and then one updates openjdk to a new version with #139 included, and then deactivates, then `$JAVA_HOME` would not be unset, but set to an empty string. I don't think this is problematic.

I don't know why I care at all about this. I fixed my own problem simply by not using `set -u` and just starting a new shell if the environment gets corrupted. Maybe this is just one tiny step closer to a perfect world, and one less open issue on github.
